packages/ethpm-v3/lib/utils.js

```diff
@@ -0,0 +1,301 @@
+const { parseTruffleArtifacts } = require("ethpm/utils/truffle");
+const Web3 = require("web3");
+const { isAddress } = require("web3-utils");
+const path = require("path");
+const fs = require("fs");
+const ENS = require("ethereum-ens");
+
+const SUPPORTED_CHAIN_IDS = {
+  1: "mainnet",
+  3: "ropsten",
+  4: "rinkeby",
+  5: "goerli",
+  42: "kovan"
+};
+
+const SUPPORTED_GENESIS_BLOCKS = {
+  "d4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3": 1,
+  "41941023680923e0fe4d74a34bdac8141f2540e3ae90623718e47d66d1ca4a2d": 3,
+  "6341fd3daf94b748c72ced5a5b26028f2474f5f00d824504e4fa37a75767e177": 4,
+  "bf7e331f7f7c1dd2e05159666b3bf8bc7a8a3a9eb1d518969eab529dd9b88c1a": 5,
+  "a3c565fc15c7478862d50ccd6561e3c06b24cc509bf388941c25ea985ce32cb9": 42
+};
+
+// Returns a list of publishable artifacts
+// aka contractTypes and deployments found in all artifacts
+async function getPublishableArtifacts(options, ethpm) {
+  var files = fs.readdirSync(options.contracts_build_directory);
+  files = files.filter(file => file.includes(".json"));
+
+  if (!files.length) {
+    var msg = `Could not locate any publishable artifacts in ${options.contracts_build_directory}. Run "truffle compile" before publishing.`;
+    return new TruffleError(msg);
+  }
+
+  const detectedNetworkIdsToBlockchainUris = {};
+  const detectedArtifacts = {};
+  const sourcePaths = {};
+
+  // group artifacts by network id to ensure consistency when converting networkId => blockchainUri
+  for (let file of files) {
+    const fileContents = JSON.parse(
+      fs.readFileSync(
+        path.join(options.contracts_build_directory, file),
+        "utf8"
+      )
+    );
+    sourcePaths[fileContents.sourcePath] = fileContents.source;
+    detectedArtifacts[fileContents.sourcePath] = fileContents;
+    const allDeploymentNetworkIds = Object.keys(fileContents.networks);
+    allDeploymentNetworkIds.forEach(deploymentNetworkId => {
+      detectedNetworkIdsToBlockchainUris[deploymentNetworkId] = "";
+    });
+  }
+
+  const requiredNetworkIds = Object.keys(detectedNetworkIdsToBlockchainUris);
+  const availableNetworkIds = Object.keys(options.networks).map(n =>
+    options.networks[n].network_id.toString()
+  );
+  for (let networkId of requiredNetworkIds) {
+    if (!availableNetworkIds.includes(networkId)) {
+      throw new TruffleError(
+        `Detected a deployment for network id: ${networkId}, but missing a provider for this network. Please make sure there is an available provider for this network id in your truffle-config.`
+      );
+    }
+  }
+
+  // generate blockchainURI for each detected networkId
+  for (let networkId of requiredNetworkIds) {
+    let targetProvider;
+    for (let provider of Object.keys(options.networks)) {
+      if (options.networks[provider].network_id == networkId) {
+        // handle ganache provider from tests
+        if (
+          options.networks[provider].provider.constructor.name == "Function"
```

Comment:
It is probably better to use `===` which is strict equals.

You could also do 
```
if (typeof options.networks[provider].provider === "function") {
```
if you like but I suppose your way is good too. Actually, how is this indexed by `provider`? Is that really some provider value?

Reply:
I've updated it to your way - this was a bit of a hack to handle ganache provider instances in the tests, while also supporting users setting provider methods as provider in `truffle-config`. Though (as in the other provider-related comment) I could be misunderstanding how users typically set their provider in `truffle-config`.